### ISSUE TITLE
zfs-local-dataset does not correct paths for chroot

### ISSUE THREAD

**Issue by hatf0** [democratic-csi/democratic-csi#178]
I'm using the `zfs-local-dataset` driver on Nomad (following some of the instructions in #164), but when trying to spin up a job, it immediately fails with the error:
```
# can safely ignore this line
/usr/local/bin/mount: illegal option -- o
filesystem 'tank/nomad/data/postgresql_db_volume' cannot be mounted at '/csi/staging/postgresql/rw-file-system-single-node-writer' due to canonicalization error: No such file or directory
```

I dug into this some more, and found that this is an issue with the plugin, due to this line:
https://github.com/democratic-csi/democratic-csi/blob/19197f9515da57f569c670d14fefb1bfd4e43460/docker/mount#L32

Nomad seems to be following spec here, as the `staging_target_path` passed *is* a valid path within the container, but democratic-csi proceeds to evaluate that path *outside* of the container, in the context of the host (the directory in the container is `/csi/staging/postgresql/rw-file-system-single-node-writer`, but the directory in the host is `/var/lib/nomad/client/csi/monolith/org.democratic-csi.zfs-dataset/staging/postgresql/rw-file-system-single-node-writer`)

**Comment by hatf0** [comment]
I played with this for a little bit, and found that actually commenting this line
https://github.com/democratic-csi/democratic-csi/blob/19197f9515da57f569c670d14fefb1bfd4e43460/docker/mount#L23
fixed things. I'm not sure if this is a fix that works everywhere, but I can say that I now have `zfs-local-dataset` working with Nomad, so....

**Comment by travisghansen** [comment]
Welcome! Interesting that it works at all on Nomad actually! That's great news.

Can you send the debug logs so I can see what the grpc request looks like? `zfs-local` is a bit unique so I want to make sure I fully understand what's going on before jumping into any solutions.

**Comment by hatf0** [comment]
> Welcome! Interesting that it works at all on Nomad actually! That's great news.
> 
> Can you send the debug logs so I can see what the grpc request looks like? `zfs-local` is a bit unique so I want to make sure I fully understand what's going on before jumping into any solutions.

`zfs-local-*` actually seems to work great otherwise, this bug not-with-standing. `zfs-local-zvol` has no issues whatsoever, but I'm (personally) not a huge fan of running ext4 on top of ZFS. 

Here's the broken log: [debug-broken.log](https://github.com/democratic-csi/democratic-csi/files/8499021/debug-broken.log)
Here's what it's like with the fix applied: [debug-good.log](https://github.com/democratic-csi/democratic-csi/files/8499034/debug-good.log)
<details>
<summary>zfs-storage-controller-dataset.nomad</summary>

```nomad
job "zfs-storage-controller-dataset" {
  datacenters = ["dc1"]
  type        = "service"

  group "controller" {
    task "controller-sidecar" {
      lifecycle {
        hook = "poststart"
        sidecar = true
      }

      driver = "docker"

      env {
        BIND_TO="unix:///csi/csi.sock"
        PROXY_TO="unix:///csi/csi.sock.internal"
      }

      config {
        image = "docker.io/democraticcsi/csi-grpc-proxy:latest"
        
        privileged = true
        ipc_mode = "host"

        # This probably isn't a good idea!
        mount {
          type = "bind"
          target = "/csi"
          # Will change based on the csi_plugin.id and your data_dir!
          source = "/var/lib/nomad/client/csi/monolith/org.democratic-csi.zfs-dataset"
          readonly = false
          bind_options {
            propagation = "rshared"
          }
        }
      }
    }

    task "controller" {
      driver = "docker"

      env {
        CSI_NODE_ID = "${attr.unique.hostname}"
      }

      csi_plugin {
        # must match --csi-name arg
        id        = "org.democratic-csi.zfs-dataset"
        type      = "monolith"
        mount_dir = "/csi"
      }

      config {
        image = "ghcr.io/hatf0/democratic-csi:latest"

        args = [
          "--csi-version=1.5.0",
          "--csi-name=org.democratic-csi.zfs-dataset",
          "--driver-config-file=${NOMAD_TASK_DIR}/driver-config-file.yaml",
          "--log-level=debug",
          "--csi-mode=controller",
          "--csi-mode=node",
          "--server-socket=/csi/csi.sock.internal",
        ]

        privileged = true
        ipc_mode = "host"
        network_mode = "host"

        mount {
          type = "bind"
          target = "/host"
          source = "/"
          readonly = false
        }
      }

      template {
        destination = "${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
driver: zfs-local-dataset

zfs:
  datasetParentName: tank/nomad/data 
  detachedSnapshotsDatasetParentName: tank/nomad/snapshots

  datasetProperties:
    # key: value

  datasetEnableQuotas: true
  datasetEnableReservation: false 
EOH
      }

      resources {
        cpu    = 250
        memory = 256
      }
    }
  }
}
```
</details>

From the logs, it looks like this issue is pretty solely rooted in the fact that the mount / umount binaries chroot away. I tried fixing this a different way (see [here](https://github.com/hatf0/democratic-csi)), but I quickly ran into weird mount propagation issues and gave up.

**Comment by travisghansen** [comment]
There must be a disconnect somewhere. `staging_target_path` isn't supposed to be a path in the container. Or said slightly differently, whatever the base path is for the staging/publish mounts should be bind mounted/shared between the host and the container...and use the same path for both (ie: `/var/lib/nomad` on the host should be mounted to `/var/lib/nomad` in the container).

Do you have `Nomad` logs for the workflow using a non-local driver?

**Comment by travisghansen** [comment]
Actually my comment may be incorrect. That is simply convention and the spec seems to indicate otherwise. Let me think on this one a little more about an appropriate way to solve.

I may have some follow up questions about your setup however.

**Comment by hatf0** [comment]
> Do you have `Nomad` logs for the workflow using a non-local driver?

I do not, unfortunately. I briefly tried playing around with non-local drivers, but ran into some blockers as the applications I'm bringing up are not huge fans of iSCSI / NFS. I can't remember if there were any other issues outside of application-level ones, though.

> Actually my comment may be incorrect. That is simply convention and the spec seems to indicate otherwise. Let me think on this one a little more about an appropriate way to solve.
> 
> I may have some follow up questions about your setup however.

No worries -- feel free to ask away and I'll try to provide any information best I can. I'd really like to get this bug squashed, as it's one of the main blockers preventing me from deploying the plugin in production. 

Just a heads up too: our Nomad setup is a *smidge* more complex then your typical setup as well. In the name of security, basically no job (except for the CSI plugin host) runs on Docker -- everything uses the containerd backend, which then in turn uses ﻿[Kata Containers](https://katacontainers.io) so that every container is actually a full fledged light-weight virtual machine running under [cloud-hypervisor](https://github.com/cloud-hypervisor/cloud-hypervisor). This is fairly complicated (unfortunately), but I can say that with my minor fix, everything works just fine.

**Comment by travisghansen** [comment]
Great! I’m familiar with kata and that’s a pretty cool setup.

What I’ll do is make that value respect what’s set by the user if set. Then you can just deploy with the env var set to 0 and it will work.

I was (apparently) erroneously under the impression the container would need the zfs tools/etc to be able to mount zfs file systems. I suppose the required bits must be in the kernel and no userspace tools are required. It’s a simple fix and I’ll get a fix into the `next` image for you to test.

**Comment by hatf0** [comment]
> Great! I’m familiar with kata and that’s a pretty cool setup.
> 
> What I’ll do is make that value respect what’s set by the user if set. Then you can just deploy with the env var set to 0 and it will work.
> 
> I was (apparently) erroneously under the impression the container would need the zfs tools/etc to be able to mount zfs file systems. I suppose the required bits must be in the kernel and no userspace tools are required. It’s a simple fix and I’ll get a fix into the `next` image for you to test.

Sounds good, let me know when it's ready to test and I'll deploy it to our test cluster.

**Comment by travisghansen** [comment]
Give it a try with the most recent `next` images. Be sure to set the env var `USE_HOST_MOUNT_TOOLS=0`

**Comment by travisghansen** [comment]
Any better?

**Comment by hatf0** [comment]
Sorry - I haven't been able to make the time to test this change out yet. I'll give it a run today and see how it is.

**Comment by travisghansen** [comment]
How'd it go?

**Comment by hatf0** [comment]
Sorry about the huge delay there, I tore down my Nomad cluster and forgot to test this. I just set it back up, and it seems to work fine with `USE_HOST_MOUNT_TOOLS=0` with this job configuration:
```
job "zfs-storage-controller-dataset" {
  datacenters = ["dc1"]
  type        = "service"

  constraint {
    attribute = "${node.unique.id}"
    value = "7114b0d2-c3a0-2dc7-e15d-aa8a63d9ffda"
  }

  group "controller" {
    task "controller-sidecar" {
      lifecycle {
        hook = "poststart"
        sidecar = true
      }

      driver = "docker"

      env {
        BIND_TO="unix:///csi/csi.sock"
        PROXY_TO="unix:///csi/csi.sock.internal"
      }

      config {
        image = "docker.io/democraticcsi/csi-grpc-proxy:latest"
        
        privileged = true
        ipc_mode = "host"

        # This probably isn't a good idea!
        mount {
          type = "bind"
          target = "/csi"
          # Will change based on the csi_plugin.id and your data_dir!
          source = "/var/lib/nomad/client/csi/monolith/org.democratic-csi.zfs-dataset"
          readonly = false
          bind_options {
            propagation = "rshared"
          }
        }
      }
    }

    task "controller" {
      driver = "docker"

      env {
        CSI_NODE_ID = "${attr.unique.hostname}"
        USE_HOST_MOUNT_TOOLS = "0"
      }

      csi_plugin {
        # must match --csi-name arg
        id        = "org.democratic-csi.zfs-dataset"
        type      = "monolith"
        mount_dir = "/csi"
      }

      config {
        image = "democraticcsi/democratic-csi:next"

        args = [
          "--csi-version=1.5.0",
          "--csi-name=org.democratic-csi.zfs-dataset",
          "--driver-config-file=${NOMAD_TASK_DIR}/driver-config-file.yaml",
          "--log-level=debug",
          "--csi-mode=controller",
          "--csi-mode=node",
          "--server-socket=/csi/csi.sock.internal",
        ]

        privileged = true
        ipc_mode = "host"
        network_mode = "host"

        mount {
          type = "bind"
          target = "/host"
          source = "/"
          readonly = false
        }
      }

      template {
        destination = "${NOMAD_TASK_DIR}/driver-config-file.yaml"

        data = <<EOH
driver: zfs-local-dataset

zfs:
  datasetParentName: tank/nomad/data 
  detachedSnapshotsDatasetParentName: tank/nomad/snapshots

  datasetProperties:
    # key: value

  datasetEnableQuotas: true
  datasetEnableReservation: false 
  datasetPermissionsMode: "0777"
EOH
      }

      resources {
        cpu    = 250
        memory = 256
      }
    }
  }
}
```
The bind mount to `/host` is still necessary, but otherwise it seems to work great.

**Comment by travisghansen** [comment]
Ok great!

**Comment by travisghansen** [comment]
Fixed in commit: https://github.com/democratic-csi/democratic-csi/commit/8ac44c0a67ddf9f59476b8f04b710c364a063e3c

Released in `v1.7.0`
